File: terminalGraph.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <windows.h>
#include <sstream>

class Config {
public:
    bool is2D;
    std::string dataFile;
    int windowSize;

    Config() : is2D(false), windowSize(100) {}
};

Config parseArgs(int argc, char* argv[]) {
    Config config;
    int i = 1;
    while (i < argc) {
        std::string arg = argv[i];
        if (arg == "-2d") {
            config.is2D = true;
        } else if (arg == "-f" && i + 1 < argc) {
            config.dataFile = argv[i + 1];
            i++;
        } else if (arg == "-w" && i + 1 < argc) {
            config.windowSize = std::stoi(argv[i + 1]);
            i++;
        }
        i++;
    }
    return config;
}

void calculateRange(const std::vector<double>& data, double& minVal, double& maxVal, double& rangeVal) {
    minVal = *std::min_element(data.begin(), data.end());
    maxVal = *std::max_element(data.begin(), data.end());
    rangeVal = maxVal - minVal;
    if (rangeVal == 0) {
        rangeVal = 1; // Avoid division by zero
    }
}

void clearScreen() {
    system("cls");
}

void updateWaveform(const std::vector<double>& data, int width, int height, double minVal, double maxVal, double rangeVal, int currentIndex) {
    // Clear the screen
    clearScreen();

    // Draw header
    std::cout << "=======================================\n";
    std::cout << "Real-time Waveform Display\n";
    std::cout << "Window Size: " << data.size() << "\n";
    std::cout << "Range: " << minVal << " to " << maxVal << "\n";
    std::cout << "=======================================\n";

    // Calculate waveform points positions
    std::vector<std::pair<int, int>> waveformPoints;
    for (size_t i = 0; i < data.size(); i++) {
        double normalized = (data[i] - minVal) / rangeVal;
        int y = height - 1 - static_cast<int>(normalized * (height - 2));
        // Ensure y is within bounds
        y = std::max<int>(0, std::min<int>(y, height - 1));
        int x;
        if (data.size() > 1) {
            x = static_cast<int>((i * (width - 10)) / (data.size() - 1));
        } else {
            x = 0;
        }
        // Ensure x is within bounds
        x = std::max<int>(0, std::min<int>(x, width - 11));
        waveformPoints.emplace_back(y, x);
    }

    // Create a 2D grid for the waveform
    std::vector<std::vector<char>> grid(height, std::vector<char>(width - 10, ' '));

    // Draw waveform points
    for (const auto& point : waveformPoints) {
        int y = point.first;
        int x = point.second;
        if (y >= 0 && y < height && x >= 0 && x < width - 10) {
            grid[y][x] = '*';
        }
    }

    // Draw simple lines between consecutive points to make the waveform more continuous
    for (size_t i = 1; i < waveformPoints.size(); i++) {
        int prevY = waveformPoints[i - 1].first;
        int prevX = waveformPoints[i - 1].second;
        int currY = waveformPoints[i].first;
        int currX = waveformPoints[i].second;

        // Only draw lines if the points are close enough to avoid cluttering
        if (std::abs(currY - prevY) <= 2 && std::abs(currX - prevX) <= 2) {
            // Draw horizontal line
            if (currY == prevY) {
                int startX = std::min<int>(prevX, currX);
                int endX = std::max<int>(prevX, currX);
                for (int x = startX; x <= endX; x++) {
                    if (x >= 0 && x < width - 10) {
                        grid[currY][x] = '*';
                    }
                }
            }
            // Draw vertical line
            else if (currX == prevX) {
                int startY = std::min<int>(prevY, currY);
                int endY = std::max<int>(prevY, currY);
                for (int y = startY; y <= endY; y++) {
                    if (y >= 0 && y < height) {
                        grid[y][currX] = '*';
                    }
                }
            }
        }
    }

    // Draw Y-axis with waveform
    for (int y = 0; y < height; y++) {
        // Calculate the Y-axis value
        double val;
        if (y == 0) {
            val = maxVal;
        } else if (y == height - 1) {
            val = minVal;
        } else {
            val = maxVal - (y * rangeVal) / (height - 2);
        }

        // Print Y-axis label
        std::cout << std::fixed;
        std::cout.precision(2);
        std::cout.width(9);
        std::cout << val << " |";

        // Print the waveform for this line
        for (int x = 0; x < width - 10; x++) {
            std::cout << grid[y][x];
        }
        std::cout << "\n";
    }

    // Draw X-axis labels
    std::cout << std::string(10, ' ');
    for (int i = 0; i < 5; i++) {
        int label = i * (data.size() - 1) / 4;
        std::stringstream ss;
        ss << label;
        std::string labelStr = ss.str();
        int padding = (width - 10) / 4;
        int leftPad = (padding - labelStr.length()) / 2;
        int rightPad = padding - leftPad - labelStr.length();
        std::cout << std::string(leftPad, ' ') << labelStr << std::string(rightPad, ' ');
    }
    std::cout << "\n";

    // Draw unit
    std::cout << "Unit: normalized\n";
    std::cout << "=======================================\n";
    std::cout << "Press Ctrl+C to exit\n";
    std::cout << "=======================================\n";
    std::cout << "Current Index: " << currentIndex << "\n";
}

int main(int argc, char* argv[]) {
    try {
        Config config = parseArgs(argc, argv);
        if (!config.is2D || config.dataFile.empty()) {
            std::cout << "Usage: terminalGraph -2d -f dataFile [-w windowSize]\n";
            return 1;
        }

        // First pass: read all data to calculate range
        std::vector<double> allData;
        try {
            std::ifstream file(config.dataFile);
            if (!file.is_open()) {
                std::cout << "Error: Could not open file\n";
                return 1;
            }

            std::string line;
            while (std::getline(file, line)) {
                // Trim whitespace
                size_t start = line.find_first_not_of(" \t");
                size_t end = line.find_last_not_of(" \t");
                if (start != std::string::npos && end != std::string::npos) {
                    line = line.substr(start, end - start + 1);
                } else {
                    continue; // Skip empty lines
                }

                try {
                    double value = std::stod(line);
                    allData.push_back(value);
                } catch (const std::invalid_argument&) {
                    continue; // Skip non-numeric lines
                }
            }
            file.close();
        } catch (const std::exception& e) {
            std::cout << "Error reading file: " << e.what() << "\n";
            return 1;
        }

        if (allData.empty()) {
            std::cout << "Error: No valid data found in file\n";
            return 1;
        }

        // Calculate range for fixed scale
        double minVal, maxVal, rangeVal;
        calculateRange(allData, minVal, maxVal, rangeVal);

        // Terminal dimensions
        int width = 80;
        int height = 20;

        // Display data in real-time with sliding window
        std::vector<double> windowData;
        int totalDataCount = 0;

        try {
            std::ifstream file(config.dataFile);
            if (!file.is_open()) {
                std::cout << "Error: Could not open file\n";
                return 1;
            }

            std::string line;
            while (std::getline(file, line)) {
                // Trim whitespace
                size_t start = line.find_first_not_of(" \t");
                size_t end = line.find_last_not_of(" \t");
                if (start != std::string::npos && end != std::string::npos) {
                    line = line.substr(start, end - start + 1);
                } else {
                    continue; // Skip empty lines
                }

                try {
                    double value = std::stod(line);
                    // Add new data to the window
                    windowData.push_back(value);
                    totalDataCount++;

                    // If window exceeds size, remove the oldest data
                    if (windowData.size() > config.windowSize) {
                        windowData.erase(windowData.begin());
                    }

                    // Update the waveform
                    updateWaveform(windowData, width, height, minVal, maxVal, rangeVal, totalDataCount);

                    // Add a small delay to simulate real-time data acquisition
                    Sleep(30);

                } catch (const std::invalid_argument&) {
                    continue; // Skip non-numeric lines
                }
            }
            file.close();
        } catch (const std::exception& e) {
            std::cout << "Error reading file: " << e.what() << "\n";
            return 1;
        }

    } catch (const std::exception& e) {
        std::cout << "Error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: terminalGraph.py
import sys
import time

class Config:
    def __init__(self):
        self.is2D = False
        self.dataFile = ""
        self.windowSize = 100

def parseArgs(argv):
    config = Config()
    i = 1
    while i < len(argv):
        arg = argv[i]
        if arg == "-2d":
            config.is2D = True
        elif arg == "-f" and i + 1 < len(argv):
            config.dataFile = argv[i + 1]
            i += 1
        elif arg == "-w" and i + 1 < len(argv):
            config.windowSize = int(argv[i + 1])
            i += 1
        i += 1
    return config

def calculateRange(data):
    """Calculate the min and max values for the entire dataset"""
    minVal = min(data)
    maxVal = max(data)
    rangeVal = maxVal - minVal
    if rangeVal == 0:
        rangeVal = 1  # Avoid division by zero
    return minVal, maxVal, rangeVal

def clearScreen():
    """Clear the terminal screen"""
    # Use a more compatible way to clear the screen
    import os
    os.system('cls' if os.name == 'nt' else 'clear')

def drawInitialScreen(width, height, minVal, maxVal, rangeVal, windowSize):
    """Draw the initial screen with axes"""
    # Clear screen
    clearScreen()
    
    # Draw header
    print("=======================================")
    print("Real-time Waveform Display")
    print(f"Window Size: {windowSize}")
    print(f"Range: {minVal:.2f} to {maxVal:.2f}")
    print("=======================================")
    
    # Draw scale on y-axis
    print(f"{maxVal:9.2f} |")
    for y in range(1, height - 1):
        val = maxVal - (y * rangeVal) / (height - 2)
        print(f"{val:9.2f} |" + " " * (width - 10))
    print(f"{minVal:9.2f} |" + "-" * (width - 10))
    
    # Draw x-axis labels
    print(" " * 10, end="")
    for i in range(5):
        label = i * (windowSize - 1) // 4
        print(f"{label:^{ (width - 10) // 4 }}", end="")
    print()
    
    # Draw unit
    print("Unit: normalized")
    print("=======================================")
    print("Press Ctrl+C to exit")
    print("=======================================")
    print("Current Index: 0")

def updateWaveform(data, width, height, minVal, maxVal, rangeVal, currentIndex):
    """Update only the waveform points"""
    # Clear the screen and redraw everything (more reliable approach)
    clearScreen()
    
    # Redraw header
    print("=======================================")
    print("Real-time Waveform Display")
    print(f"Window Size: {len(data)}")
    print(f"Range: {minVal:.2f} to {maxVal:.2f}")
    print("=======================================")
    
    # Calculate waveform points positions
    waveform_points = []
    for i in range(len(data)):
        normalized = (data[i] - minVal) / rangeVal
        y = height - 1 - int(normalized * (height - 2))
        # Ensure y is within bounds
        y = max(0, min(y, height - 1))
        if len(data) > 1:
            x = int((i * (width - 10)) / (len(data) - 1))
        else:
            x = 0
        # Ensure x is within bounds
        x = max(0, min(x, width - 11))
        waveform_points.append((y, x))
    
    # Create a 2D grid for the waveform
    grid = [[" " for _ in range(width - 10)] for _ in range(height)]
    
    # Draw waveform points
    for y, x in waveform_points:
        if 0 <= y < height and 0 <= x < width - 10:
            grid[y][x] = "*"
    
    # Draw simple lines between consecutive points to make the waveform more continuous
    for i in range(1, len(waveform_points)):
        prev_y, prev_x = waveform_points[i-1]
        curr_y, curr_x = waveform_points[i]
        
        # Only draw lines if the points are close enough to avoid cluttering
        if abs(curr_y - prev_y) <= 2 and abs(curr_x - prev_x) <= 2:
            # Draw horizontal line
            if curr_y == prev_y:
                start_x = min(prev_x, curr_x)
                end_x = max(prev_x, curr_x)
                for x_line in range(start_x, end_x + 1):
                    if 0 <= x_line < width - 10:
                        grid[curr_y][x_line] = "*"
            # Draw vertical line
            elif curr_x == prev_x:
                start_y = min(prev_y, curr_y)
                end_y = max(prev_y, curr_y)
                for y_line in range(start_y, end_y + 1):
                    if 0 <= y_line < height:
                        grid[y_line][curr_x] = "*"
    
    # Redraw Y-axis with waveform
    for y in range(height):
        # Calculate the Y-axis value
        if y == 0:
            val = maxVal
        elif y == height - 1:
            val = minVal
        else:
            val = maxVal - (y * rangeVal) / (height - 2)
        
        # Print Y-axis label
        print(f"{val:9.2f} |", end="")
        
        # Print the waveform for this line
        print("" .join(grid[y]), end="")
        print()
    
    # Redraw X-axis labels
    print(" " * 10, end="")
    for i in range(5):
        label = i * (len(data) - 1) // 4
        print(f"{label:^{ (width - 10) // 4 }}", end="")
    print()
    
    # Redraw unit
    print("Unit: normalized")
    print("=======================================")
    print("Press Ctrl+C to exit")
    print("=======================================")
    print(f"Current Index: {currentIndex}")

def main():
    try:
        config = parseArgs(sys.argv)
        if not config.is2D or not config.dataFile:
            print("Usage: terminalGraph.py -2d -f dataFile [-w windowSize]")
            return 1
        
        # First pass: read all data to calculate range
        allData = []
        try:
            with open(config.dataFile, 'r') as file:
                for line in file:
                    line = line.strip()
                    if line:
                        try:
                            value = float(line)
                            allData.append(value)
                        except ValueError:
                            continue
        except Exception as e:
            print(f"Error reading file: {e}")
            return 1
        
        if not allData:
            print("Error: No valid data found in file")
            return 1
        
        # Calculate range for fixed scale
        minVal, maxVal, rangeVal = calculateRange(allData)
        
        # Terminal dimensions
        width = 80
        height = 20
        
        # Display data in real-time with sliding window
        windowData = []
        totalDataCount = 0
        
        try:
            with open(config.dataFile, 'r') as file:
                for line in file:
                    line = line.strip()
                    if line:
                        try:
                            value = float(line)
                            # Add new data to the window
                            windowData.append(value)
                            totalDataCount += 1
                            
                            # If window exceeds size, remove the oldest data
                            if len(windowData) > config.windowSize:
                                windowData.pop(0)
                            
                            # Update the waveform
                            updateWaveform(windowData, width, height, minVal, maxVal, rangeVal, totalDataCount)
                            
                            # Add a small delay to simulate real-time data acquisition
                            time.sleep(0.03)
                            
                        except ValueError:
                            continue
        except Exception as e:
            print(f"Error reading file: {e}")
            return 1
    
    except Exception as e:
        print(f"Error: {e}")
        return 1
    
    return 0

if __name__ == "__main__":
    main()